File: game/src/main/java/com/exjets/aetheria/npc/Npc.java
package com.exjets.aetheria.npc;

import com.exjets.aetheria.world.Position;

/** A spawned creature: where it stands, how hurt it is and what it is currently doing. */
public class Npc {

    private final NpcDef def;
    private final Position spawn;
    private final int wanderRadius;

    private Position position;
    private int hitpoints;
    private int respawnTimer;
    private int attackCooldown;
    private boolean inCombat;
    private int lastDamage = -1;
    private int hitsplatTimer;

    public Npc(NpcDef def, Position spawn, int wanderRadius) {
        this.def = def;
        this.spawn = spawn;
        this.position = spawn;
        this.wanderRadius = wanderRadius;
        this.hitpoints = def.hitpoints();
    }

    public NpcDef def() {
        return def;
    }

    public String name() {
        return def.name();
    }

    public Position spawn() {
        return spawn;
    }

    public int wanderRadius() {
        return wanderRadius;
    }

    public Position position() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public int hitpoints() {
        return hitpoints;
    }

    public int maxHitpoints() {
        return def.hitpoints();
    }

    public void damage(int amount) {
        hitpoints = Math.max(0, hitpoints - amount);
        lastDamage = amount;
        hitsplatTimer = 3;
    }

    public boolean isDead() {
        return hitpoints <= 0;
    }

    /** Dead creatures stay out of the world until their respawn timer elapses. */
    public boolean isVisible() {
        return respawnTimer <= 0;
    }

    public void kill() {
        respawnTimer = def.respawnTicks();
        inCombat = false;
        attackCooldown = 0;
    }

    public void tickTimers() {
        if (respawnTimer > 0 && --respawnTimer == 0) {
            hitpoints = def.hitpoints();
            position = spawn;
        }
        if (attackCooldown > 0) {
            attackCooldown--;
        }
        if (hitsplatTimer > 0 && --hitsplatTimer == 0) {
            lastDamage = -1;
        }
    }

    public boolean canAttackNow() {
        return attackCooldown <= 0;
    }

    public void startAttackCooldown(int ticks) {
        attackCooldown = ticks;
    }

    public boolean isInCombat() {
        return inCombat;
    }

    public void setInCombat(boolean inCombat) {
        this.inCombat = inCombat;
    }

    /** The damage number floating over the creature, or -1 when there is none. */
    public int lastDamage() {
        return lastDamage;
    }

    public boolean isWithinWanderRange(Position candidate) {
        return candidate.distanceTo(spawn) <= Math.max(wanderRadius, 1);
    }
}

File: src/game/npc/npc.ts
import { NpcDef } from "./defs";
import { Position, distance } from "../world/position";

/** A spawned creature: where it stands, how hurt it is and what it is doing. */
export class Npc {
  position: Position;
  private hp: number;
  private respawnTimer = 0;
  private attackCooldown = 0;
  private fighting = false;
  private damageShown = -1;
  private hitsplatTimer = 0;

  constructor(readonly def: NpcDef, readonly spawn: Position, readonly wanderRadius: number) {
    this.position = spawn;
    this.hp = def.hitpoints;
  }

  get name(): string {
    return this.def.name;
  }

  hitpoints(): number {
    return this.hp;
  }

  maxHitpoints(): number {
    return this.def.hitpoints;
  }

  damage(amount: number): void {
    this.hp = Math.max(0, this.hp - amount);
    this.damageShown = amount;
    this.hitsplatTimer = 3;
  }

  isDead(): boolean {
    return this.hp <= 0;
  }

  /** Dead creatures stay out of the world until their respawn timer elapses. */
  isVisible(): boolean {
    return this.respawnTimer <= 0;
  }

  kill(): void {
    this.respawnTimer = this.def.respawnTicks;
    this.fighting = false;
    this.attackCooldown = 0;
  }

  tickTimers(): void {
    if (this.respawnTimer > 0 && --this.respawnTimer === 0) {
      this.hp = this.def.hitpoints;
      this.position = this.spawn;
    }
    if (this.attackCooldown > 0) this.attackCooldown--;
    if (this.hitsplatTimer > 0 && --this.hitsplatTimer === 0) this.damageShown = -1;
  }

  canAttackNow(): boolean {
    return this.attackCooldown <= 0;
  }

  startAttackCooldown(ticks: number): void {
    this.attackCooldown = ticks;
  }

  isInCombat(): boolean {
    return this.fighting;
  }

  setInCombat(fighting: boolean): void {
    this.fighting = fighting;
  }

  /** The damage number floating over the creature, or -1 when there is none. */
  lastDamage(): number {
    return this.damageShown;
  }

  isWithinWanderRange(candidate: Position): boolean {
    return distance(candidate, this.spawn) <= Math.max(this.wanderRadius, 1);
  }
}
